pkg/api/v1/schema.go

```diff
@@ -19,346 +19,474 @@
+	Resources ResourceRequirements `json:"resources,omitempty"`
+	// Firmware
+	// +optional
+	Firmware *Firmware `json:"firmware,omitempty"`
+	// Clock sets the clock and timers of the vm.
+	// +optional
+	Clock *Clock `json:"clock,omitempty"`
+	// Features like acpi, apic, hyperv
+	// +optional
+	Features *Features `json:"features,omitempty"`
+	// Devices allows adding disks, network interfaces, ...
+	Devices Devices `json:"devices"`
+}
+
+type ResourceRequirements struct {
+	// Initial is a description of the initial vm resources.
+	// Valid resource keys are "memory" and "cpu".
+	// +optional
+	Initial v1.ResourceList `json:"initial,omitempty"`
+}
+
+type Firmware struct {
+	// UID reported by the vm bios
+	// Defaults to a random generated uid
+	UID types.UID `json:"uid,omitempty"`
 }
 
 type Devices struct {
-	Emulator   string      `json:"emulator,omitempty"`
-	Interfaces []Interface `json:"interfaces,omitempty"`
-	Channels   []Channel   `json:"channels,omitempty"`
-	Video      []Video     `json:"video,omitempty"`
-	Graphics   []Graphics  `json:"graphics,omitempty"`
-	Ballooning *Ballooning `json:"memballoon,omitempty"`
 	Disks      []Disk      `json:"disks,omitempty"`
-	Serials    []Serial    `json:"serials,omitempty"`
-	Consoles   []Console   `json:"consoles,omitempty"`
+	Interfaces []Interface `json:"interfaces,omitempty"`
 	Watchdog   *Watchdog   `json:"watchdog,omitempty"`
 }
 
-// BEGIN Disk -----------------------------
-
 type Disk struct {
-	Device    string         `json:"device,omitempty"`
-	Snapshot  string         `json:"snapshot,omitempty"`
-	Type      string         `json:"type"`
-	Source    DiskSource     `json:"source,omitempty"`
-	Target    DiskTarget     `json:"target"`
-	Serial    string         `json:"serial,omitempty"`
-	Driver    *DiskDriver    `json:"driver,omitempty"`
-	ReadOnly  *ReadOnly      `json:"readOnly,omitempty"`
-	Auth      *DiskAuth      `json:"auth,omitempty"`
-	CloudInit *CloudInitSpec `json:"cloudinit,omitempty"`
-}
-
-type DiskAuth struct {
-	Username string      `json:"username"`
-	Secret   *DiskSecret `json:"secret,omitempty"`
-}
-
-type DiskSecret struct {
-	Type  string `json:"type"`
-	Usage string `json:"usage"`
-}
-
-type ReadOnly struct{}
-
-type DiskSource struct {
-	File          string          `json:"file,omitempty"`
-	StartupPolicy string          `json:"startupPolicy,omitempty"`
-	Protocol      string          `json:"protocol,omitempty"`
-	Name          string          `json:"name,omitempty"`
-	Host          *DiskSourceHost `json:"host,omitempty"`
+	// Name is the device name
+	Name string `json:"name"`
+	// Name of the volume which is referenced
+	// Must match the Name of a Volume.
+	VolumeName string `json:"volumeName"`
+	// DiskDevice specifies as which device the disk should be added to the guest
+	// Defaults to Disk
+	DiskDevice `json:",inline"`
+}
+
+// Represents the target of a volume to mount.
+// Only one of its members may be specified.
```

Comment:
Do we expect an admission controller will enforce this at some point?

Reply:
Yes.

Reply:
reflected this in #556